Multiple SocketCAN instance support.

sumitbatra-nxp opened the pull request:
Adding support for multiple socketCAN interfaces by
making use of instances of DT_DRV_COMPAT for
network devices instead of a single network interface
Each socketCAN instance has a corresponding
DT Node named net_canbus<0>,<1>.

The net_canbus DT Node needs to be a child of the CAN DT Node.
The CAN and socketCAN devices can be accessed
by applications by parent child DT macros.
SocketCAN applications can access socketCAN node and
hence the network interface with its node id.


Tested on RDDRONE board.


Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -944,6 +944,18 @@ struct net_if *net_if_get_default(void);
  */
 struct net_if *net_if_get_first_by_type(const struct net_l2 *l2);
 
+/**
+ * @brief Get nth network interface according to its type.
+ *
+ * @param l2 Layer 2 type of the network interface.
+ *
+ * @param node_pos position of the network interface in Layer 2 type list.
+ *
+ * @return nth network interface of a given type or NULL if no such
+ * interfaces was found.
+ */
+struct net_if *net_if_get_nth_by_type(const struct net_l2 *l2, int node_pos);
+
```
I have some reservations for this function. The usecase for this seems to be very specific and depends on the interface position which can change if the interfaces are placed in the different order. Also all of this could be done in the application side which can use better heuristic to determine which interface to select.

sumitbatra-nxp replied:
Removed it as per your suggestion

bperseghetti committed:
can: Multiple SocketCAN instance support.

Adding support for multiple socketCAN interfaces by
making socketCAN driver multi-instanced.
Each socketCAN instance has a corresponding
DT Node named net_canbus<0>,<1>.

The net_canbus DT Node needs to be a child of the CAN DT Node.
The CAN and socketCAN devices can be accessed
by applications by parent child DT macros.
SocketCAN applications can access socketCAN node and
hence the network interface with its node id.

Tested on RDDRONE board.

Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Sumit Batra <sumit.batra@nxp.com>

jukkar reviewed (changes requested):
After some more thinking, I think we cannot rely on interface ordering here. The interface ordering is not guaranteed and might change for example between different zephyr release.
Would there be any other way to distinguish the canbus interface, like address or name etc, that could be used to select the network interface?

bperseghetti reviewed (commented):
@jukkar thanks for the great suggestions, please let us know if our fixes resolved all the issues.

sumitbatra-nxp commented:
> After some more thinking, I think we cannot rely on interface ordering here. The interface ordering is not guaranteed and might change for example between different zephyr release. Would there be any other way to distinguish the canbus interface, like address or name etc, that could be used to select the network interface?

>> In our latest socketcan driver changes we have bound the net_canbus<x> DT node with its parent can<x or y> or flexcan<x or y> node.
Now the author of the board dts and SoC dtsi has a free will to define any net_canbus node under any CAN node.
The applications can use socketcan node id to access the socketcan device.

jukkar reviewed (dismissed):
Thanks, this looks much better now.

henrikbrixandersen commented on boards/arm/rddrone_fmuk66/rddrone_fmuk66.dts:
```diff
@@ -35,7 +35,8 @@
 		zephyr,shell-uart = &lpuart0;
 		zephyr,uart-pipe = &lpuart0;
-		zephyr,canbus = &flexcan0;
+		zephyr,canbus0 = &flexcan0;
+		zephyr,canbus1 = &flexcan1;
```
You are introducing a new naming for chosen nodes here. This will break other subsystems and samples relying on the current naming.

Why are these new chosen nodes needed?

sumitbatra-nxp replied:
In the current Zephyr code, 
I can see some Device Trees have defined CAN nodes with node id's- **can<x>** while others have used **flexcan<x>**
In short, I there is no generic way of creating node id's for a CAN node.
Now, I wanted to create a generic application which uses multiple CAN nodes, so I thought that the application writer (based on the number of CAN PHYs on his board) could create **n** number of CHOSEN nodes like **zephyr_canbus<x>**.
I know that this will break other subsystems, and I can think more over it, but I feel that the original naming wasn't written with multiple CAN instances in mind.
@henrikbrixandersen I welcome your thoughts on this, please free to correct my understanding.

henrikbrixandersen replied:
The "node id" you refer to is called the nodelabel, which matches (or at least should match) what the given IP block is called in the SoC datasheet.

For out-of-tree applications, you can add whatever chosen nodes and/or devicetree aliases you need in a devicetree overlay. For in-tree  boards, samples, and tests we use the Zephyr-specific chosen nodes as documented here: https://docs.zephyrproject.org/latest/build/dts/api/api.html#zephyr-specific-chosen-nodes The rationale behind only supporting one CAN bus chosen node in Zephyr is that no in-tree code have yet needed to be able to support more than one CAN controller.

If you would like to propose a change to the naming of the Zephyr-specific chosen nodes, please open a separate RFC for this.

sumitbatra-nxp replied:
Just to make sure I understood correctly -
I can write any chosen node names (even which breaks the rules mentioned in the link you provided) in the overlay of my out of tree application ?

carlescufi replied:
@sumitbatra-nxp ping

sumitbatra-nxp replied:
I would take back this chosen node change, since I feel it makes more sense to choose just one instance of a subsystem (out of many)

henrikbrixandersen commented on dts/arm/nxp/nxp_k66.dtsi:
```diff
@@ -34,6 +34,10 @@
 			sample-point = <875>;
 			status = "disabled";
+			net_canbus1 {
+				compatible = "net-canbus";
```
You are introducing a new DTS compatible here without adding a DTS binding. 

DTS is supposed to describe hardware, not software. I am not fond of describing what is basically a network stack on top of CAN via the devicetree.

henrikbrixandersen replied:
I have given this some more thought and will try to come up with an alternative solution, where the CAN driver instances are placed in an iteratable section similar to what was recently done for the sensor drivers. This will allow the network canbus driver to create a network interface for each CAN controller without having to add additional configuration to the dts.

bperseghetti replied:
> 

Appreciate you looking into it!

sumitbatra-nxp replied:
@henrikbrixandersen ..
Please evaluate this idea...
If I trigger socketcan init from CAN device init (let's say I add an API to trigger socketcan init in can_common.c or can_handler.c) (this choice of creating a socketcan interface can be based on a new property in CAN DT nodes)

sumitbatra-nxp replied:
@henrikbrixandersen ..I have a few queries about putting CAN driver instances in ITERABLE_SECTION.
Q1. Can the order of CAN instances in this section vary? Since DT parsing order can change.
Q2. On what basis will network canbus driver choose its underlying CAN instance? (If this choice is based on the index of section entries, then it might change)

My point is that the correlation between the network canbus instance and the CAN instance should be consistent.

henrikbrixandersen replied:
> Q1. Can the order of CAN instances in this section vary? Since DT parsing order can change.

The instances in the iterable section will be sorted by the alphanumeric device name string.

> Q2. On what basis will network canbus driver choose its underlying CAN instance? (If this choice is based on the index of section entries, then it might change)

The canbus network driver will iterate the sorted sections in the order given by the alphanumeric device name string of the underlying CAN controller devices and create a canbus network interface for each.

igalloway commented:
@sumitbatra-nxp  @bperseghetti 
What is the summary here?
I'm trying to follow the thread, do we need to propose a change to the zephyr-specific-chosen-nodes 
https://docs.zephyrproject.org/latest/build/dts/api/api.html#zephyr-specific-chosen-nodes

in order to have multiple CAN interfaces running simultaneously (likely with differing protocols)
(See MR-CANHUBK344 board as an example hardware. There are six CAN busses)

henrikbrixandersen commented:
> do we need to propose a change to the zephyr-specific-chosen-nodes

No. I am working on a different approach.

igalloway commented:
Thanks @henrikbrixandersen 
Just in case it is relevant, note that on a board like MR-CANHUBK344 there are also different types of PHY interfaces and they could have some differing control signals. I don't think this would have an impact at this level. But i just wanted to mention it in case it impacts the software architecture.

henrikbrixandersen commented:
> Thanks @henrikbrixandersen 
> 
> Just in case it is relevant, note that on a board like MR-CANHUBK344 there are also different types of PHY interfaces and they could have some differing control signals. I don't think this would have an impact at this level. But i just wanted to mention it in case it impacts the software architecture.

Thanks. That does not impact the solution here, as GPIO-controlled CAN transceivers are already supported.

bperseghetti commented:
> > do we need to propose a change to the zephyr-specific-chosen-nodes
> 
> No. I am working on a different approach.

@henrikbrixandersen Is there a WIP branch/PR for your approach/work that you could potentially reference in the comments of this PR?

henrikbrixandersen commented:
> @henrikbrixandersen Is there a WIP branch/PR for your approach/work that you could potentially reference in the comments of this PR?

Unfortunately, the work is not yet ready for consumption/review. I have done a few refactoring PRs in the last month, which lead up to adding this. The basic idea is to have `CAN_DEVICE_DT_DEFINE()` automatically register the CAN controller device as a network interface based on whether SocketCAN is enabled in Kconfig or not (inspired by the ieee 802.15.4 drivers).

jukkar commented:
@henrikbrixandersen should we just close this instead of removing the stale label, it does not look like this is progressing in over two years?

henrikbrixandersen commented:
> @henrikbrixandersen should we just close this instead of removing the stale label, it does not look like this is progressing in over two years?
> 
> 

We could, but I still plan to address it. The solution is pending another change in the CAN driver subsystem, which I am currently working on formalizing as an RFC. 

We can move this issue to just be labeled CAN, if that helps.

jukkar commented:
No worries at all and no need to remove the net label, just wondering what the plan was.
